### ISSUE TITLE
Hosting, client ran, on a server, both?

### ISSUE THREAD

**Issue by benjaminkay93** [NaSTAuk/graphics#6]
So the final of the 6 initial questions, Do we want this to be built to run on a PC locally to kick the graphics out over CasparCG / OBS or something of the like, or in the cloud that you could host and just hit a URL, or a NaSTA hosted service that stations can just use, or a subset of all/ straight up all of them?

**Comment by TMannall** [comment]
The issue I see with a hosted solution is that it's reliant on the local PC having an internet connection. I think offering both options would be best, cloud hosting has quite a lot of benefits, like being able to build features like sharing graphics themes between stations etc.

**Comment by benjaminkay93** [comment]
sharability and discoverability will be key, depending on approach we could make it nice and straightforward for stations to find graphics that exist already, ideally stations souldn't build there own on their own githubs but would sooner opt for building it into this project for other to use, I think this could be a whole issue in itself, feel free to create an issue about this ;)

im gonna hold of my opinions as ive raised this until more people have commented, good point with the internet connect potentially being flaky, but we could get around that with service workers or other mechanisms too :)

**Comment by thomasslee97** [comment]
I agree with Tom, an internet connection can be hard to come by. But I think it should also be possible to host it on a central server and have clients connect to (such as CasparCG) - which is what we've done with the Roses graphics for this year.

Also a note: being able to host multiple instances on the same server would be nice. Matt has managed to get multiple instances of the Roses graphics running on a single server but I think that's involved a fair amount of nginx poking.

**Comment by benjaminkay93** [comment]
we could do something where we wrap each thing into a url namespace so its easy to get multiple urls on the same thing :)

**Comment by benjaminkay93** [comment]
personally, I feel both / all options are ideal but may add a bit of complexity in implementation, it would be nice to have a few more opinions too, as well as decenting ones, I can see an argument for picking just one or the other

**Comment by moschopsuk** [comment]
As an MVP running locally would be easier to devlop and test.

As for running a number of different CG "sessions" I think is worth having. We too have ran different graphics packages on a single box and it was a faff.

**Comment by benjaminkay93** [comment]
Yeah, many times that's caused issues in the past, we should consider only having one instance running that spawns different cg windows if needed?

**Comment by niklasR** [comment]
I think this is maybe related to the question I posed in #1 re architecture.

Definitely need to be able to run it locally, if only for testing.
Decoupling of template, data/business logic and gui makes sense but is that too complex?

The middle layer is probably the one that needs some decisions taken, as the rest is "just" rendered JS/HTML. Serverless or running application? A running application probably makes it easier to just set up and forget about; a serverless approach may be way cheaper to run and could be "wrapped" in a runner?

**Comment by benjaminkay93** [comment]
Maybe decoupling is abit overkill, however, it would allow us to scale to more and more components potentially easier 🤷‍♀️

The middleware is the hardest problem I see us encountering going forward, be good to hear on ideas for solving it. And tbh I haven't considered as far as serverless vs server-based approaches yet. Having a serverless view would be cheap but how would we manage the state of the views as its essentially going to be powered from on high if we go to the cloud :P
